### ISSUE TITLE
Handling owner/username in portal API calls

### ISSUE THREAD

**Issue by noahmulfinger** [Esri/arcgis-rest-js#73]
When working with user content, its common that we have to use an endpoint similar to
```
https://www.arcgis.com/sharing/rest/content/users/${username}/...
```
which requires getting the user's username. The `items` package handles this by having the user input an `owner` parameter for functions that require it. However, in the `feature-services-admin` package I've been working on, I've been assuming that the authentication is a UserSession and pulling the username from `requestOptions.authentication.username`. 

Is it always the case that *user* authentication is needed when calling an endpoint like `https://www.arcgis.com/sharing/rest/content/users/${username}/...`? If so, I think pulling from UserSession makes more sense. However, this would make `requestOptions` required, which I'm not sure matches the expected use of `requestOptions`.

I want to get people's thoughts on this so we have a consistent pattern. @patrickarlt @dbouwman @tomwayson @jgravois @araedavis 

As a side note, issues like this make me think it would be good to set up an API design reference in order to keep track of cross-package design considerations.

**Comment by jgravois** [comment]
I had the same reaction you did when i saw that `owner` was a discrete parameter in the items package.

> Is it always the case that user authentication is needed when calling an endpoint like `https://www.arcgis.com/sharing/rest/content/users/${username}/...`?

the REST documentation for _some_ item operations (like the ones pertaining to [resources](http://resources.arcgis.com/en/help/arcgis-rest-api/index.html#/Item_Resources/02r3000002r3000000/) that i mucked around with the other day in #70) state that operations can be performed by either their owner **or** organization administrators. i'll leave it to @dbouwman to confirm whether the API design aims to purposely exploit that idiosyncrasy or whether there's an opportunity to DRY them up.

either way, a guide page that discusses API design would be rad.

**Comment by patrickarlt** [comment]
I originally intended modules to override `IRequestOptions` if they needed to add additional options to a request. So we could write things like this:

```ts
import { request, IRequestOptions, IParams } from "@esri/arcgis-rest-request";
import { IItem } from "@esri/arcgis-rest-common-types"

interface AddItemToFolderOptions extends IRequestOptions {
  owner?: string;
}

export function addItemToFolder (item: IItem, folder: string, params: IParams,  requestOptions: AddItemToFolderOptions): Promise<IItem> {
  const owner = requestOptions.owner || requestOptions.session.username;
  return request(/** ... **/)
}
```

The general design for request is as follows (we should probally put this in an API design page somewhere):

```ts
interface someSugeredRequestMethodOptions extends IRequestOptions {
  someOptions?: string;
}

function someSugeredRequestMethod(
  requiredParam1: string, // this param is absolutely required to complete the request
  requiredParam2: number, // this param is also absolutely required to complete the request
  params: IParams, // any other params to pass to the rest endpoint
  options: someSugeredRequestMethodOptions // any additional options. extend IRequestOptions to add suger to request methods. This may/may not be optional and CAN override things in `params`
) {
  return request(/** ... **/)
}
```

This design lets us:

1. Specify required params in an easy way
2. Lay our own sugar on top of the APIs via `options`
3. Support any parameter in the REST API via `params`

**Comment by dbouwman** [comment]
@patrickarlt  @noahmulfinger I think using the currentUser could work for some of the items api calls - i.e. `createItem` which is clearly happening in the context of the current user.

However, calls like `removeItem` and `unprotectItem` etc, the `owner` is a required segment of the url, and (at least in hub) it's not a given (or even common) that `owner === currentUser`, so I'd rather we just make it a required parameter, vs juggling a prop on another object.

For example - when we delete an initiative or site, and we have to cascade unprotect & delete calls through a number of items (including feature-services), we batch up the requests so they run in parallel via `RSVP.allSettled([ArrayOPromises])`. In this scenario it's highly unlikely that the `currentUser` === `owner` for all the items. So, we'd need to decorate `requestOptions` with different `.owner` params for each of those... 

just feels clunky to me, and fraught with problems... like you forget to reset `requestOptions.owner` in one part of a long promise chain that is dealing with a dozen items... boom and now you have to figure out why

Anyhow - I'd argue that `required url segments === required fn params` is a reasonable, easy to follow pattern. But if everyone else on this project feels otherwise, we can always hide this detail when we wrap it into ember :)

**Comment by patrickarlt** [comment]
@dbouwman If `owner !== currentUser` you can still specify owner in my example above.

```ts
import { request, IRequestOptions, IParams } from "@esri/arcgis-rest-request";
import { IItem } from "@esri/arcgis-rest-common-types"

interface AddItemToFolderOptions extends IRequestOptions {
  owner?: string;
}

export function addItemToFolder (item: IItem, folder: string, params: IParams,  requestOptions: AddItemToFolderOptions): Promise<IItem> {
  const owner = requestOptions.owner || requestOptions.session.username;
  return request(/** ... **/)
}
```

**Comment by jgravois** [comment]
> // any additional options. extend IRequestOptions to add sugar to request methods. This may/may not be optional and CAN override things in `params`

i talked to @dbouwman about this a bit more and he could live with populating url segments like `owner` from the authentication session by default and exposing a dedicated _optional_ `owner` parameter that trumps what is in the session.

i understand the desire to make required url segments _first class citizen_ parameters instead of decorating other objects with them as properties, but from a debugging perspective, it seems just as likely to me that you'd pass in a stale `owner` argument as it would be to forget to reset `requestOptions.owner` so i see where @patrickarlt is coming from as well.

to be clear, i'd be happy with either because they both simplify the usage for more typical 'user content management scenarios'

**Comment by dbouwman** [comment]
ok - I wrote an example to demonstrate this issue --> https://gist.github.com/dbouwman/7f0133d90fe21ddc6e023315557eb2ba

In that, I search for 20 items not belonging to the current user.
I `.map` over the results to create two sets of promises... 

```js
    let promises = response.results.map((itm) => {
      requestOptions.owner = itm.owner;
      return ownerViaRequestOptions(itm, requestOptions);
    });
    let promisesWithParam = response.results.map((itm) => {
      return ownerViaParam(itm, itm.owner, requestOptions);
    });
    return Promise.all(promises.concat(promisesWithParam));
```

The promise fn's are very simple... they just check if the owner - passed either by `requestOptions.owner` or as a param, matches the `item.owner` (it should)

```js
function ownerViaRequestOptions(item, requestOptions) {
  return new Promise((resolve, request) => {
    setTimeout(() => {
      let color = 'green';
      if (item.owner !== requestOptions.owner) {
        color = 'red';
      }
      console.log(chalk[color](`REQUESTOPTIONS: Item Owner ${item.owner} and requestOptions.owner ${requestOptions.owner}`));
      resolve();
    });
  });
}

function ownerViaParam(item, owner, requestOptions) {
  return new Promise((resolve, request) => {
    setTimeout(() => {
      let color = 'green';
      if (item.owner !== owner) {
        color = 'red';
      }
      console.log(chalk[color](`PARAM: Item Owner ${item.owner} and owner param ${owner}`));
      resolve();
    });
  });
}
```
Here's the output...
![image](https://user-images.githubusercontent.com/119129/33668354-c15f2210-da5c-11e7-897f-2886366d91f9.png)

In short - don't rely on `requestOptions.<anyValue>` if you are batching requests up like this.

**Comment by patrickarlt** [comment]
@dbouwman I think your demo is a little misleading. Your issue should be resolved by changing

```ts
    let promises = response.results.map((itm) => {
      requestOptions.owner = itm.owner;
      return ownerViaRequestOptions(itm, requestOptions);
    });
```

to

```ts
let promises = response.results.map(itm => {
  return ownerViaRequestOptions(itm, {
    owner: itm.owner
  });
});
```

`requestOptions`

I agree this is a footgun but this is just like mutating a value in a `for` loop and I don't think most users are really going to be scripting things that reach across tons of users all at once. For example the developer site will always be `owner === currentUser` so we are forced to pass an additional param to every method call we make. It seems like this case is mostly Hub specific so I don't think we should let it drive our API design.

**Comment by patrickarlt** [comment]
@dbouwman the same thing could also be said for `requestOptions.portal` which is used everywhere https://github.com/Esri/arcgis-rest-js/blob/master/packages/arcgis-rest-request/src/utils/get-portal-url.ts#L16-L18

**Comment by jgravois** [comment]
in #78 we landed a change that makes `owner` an optional argument that _can_ be substituted on behalf of a user's authentication session.
